### PR TITLE
[3.7] Add support for setAsync

### PR DESCRIPTION
**Pull request by LoneRifle**
Declare and implement IMap.setAsync

Supersedes #7129 and implements #6726

### CODE REVIEW

**Comment by jerrinot** [comment]
Hello @LoneRifle,

many thanks for this useful addition! It's too late to have this included in 3.6, but I believe it's a nice improvement for 3.7. I quickly skimmed the changeset and it seems good! I'll do a proper review later. 

Once more thanks for your effort, it's much appreciated!

**Comment by cangencer** [comment]
If this will go into 3.7, the legacy-client part can be removed, there will not be legacy-client in 3.7

**Comment by LoneRifle** [comment]
I believe I was told earlier in the previous PR that this could go into 3.6.1. Would you be okay if we were to keep to this please?

**Comment by cangencer** [comment]
@LoneRifle We will discuss if this will be more suitable for 3.6.1 or 3.7, probably once 3.6 is released. So let's leave it in for now and then we can discuss further.

Overall PR looks good but is missing tests. Needs tests for

* ClientMapNearCacheTest
* NearCacheLocalInvalidationTest

I also noticed we have some duplicate tests in MapPreconditionsTest and ClientMapBasicTest, you have the precondition checks in ClientMapBasicTest which is OK I think.

**Comment by LoneRifle** [comment]
Done. Thanks for pointing out the missing tests, filled a gap missing in MapProxyImpl.

**Comment by cangencer** [comment]
run-lab-run

**Comment by hasancelik** [comment]
run-lab-run

**Comment by LoneRifle** [comment]
Rebased to reflect removal of legacy client from master. #7357 created to patch maintenance-3.x for 3.6.1.

File: hazelcast/src/main/java/com/hazelcast/core/IMap.java [hunk 1]
```diff
@@ -375,6 +375,89 @@
...
+     *
+     * @param key      the key of the map entry.
+     * @param value    the new value of the map entry.
+     * @param ttl      maximum time for this entry to stay in the map.
+     *                 0 means infinite.
+     * @param timeunit time unit for the ttl.
+     * @return Future on which to block.
+     * @throws NullPointerException if the specified key or value is null.
+     * @see java.util.concurrent.Future
+     */
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/core/IMap.java]: the params  & text reference to TTL values, which are not on this overload

**Comment by cangencer** [comment]
looks good except one minor comment regarding javadoc

File: hazelcast/src/main/java/com/hazelcast/core/IMap.java (hunk 1 above)
Comment (LoneRifle) [inline hazelcast/src/main/java/com/hazelcast/core/IMap.java]: Corrected in 9d90891

**Comment by cangencer** [comment]
:+1:

**Comment by LoneRifle** [comment]
Could someone change the milestone for this to 3.7 (as it's going to master) and apply the 3.6.1 milestone to #7357 please?

**Comment by jerrinot** [comment]
done

**Comment by jerrinot** [comment]
:+1:

**Pull request merged**